---
title: "03_universal_enrichment"
documentclass: ctexart
output:
  bookdown::pdf_book:
    latex_engine: xelatex
  html_document: default
date: "2024-05-29"
---

# Universal enrichment analysis {#universal-api}

```{r include=FALSE}
library(knitr)
opts_chunk$set(message=FALSE, warning=FALSE, eval=TRUE, echo=TRUE, cache=TRUE)
library(clusterProfiler)
```

The `clusterProfiler` package [@yu2012] supports both hypergeometric test and gene set enrichment analyses of many ontology/pathway, but it's still not enough for users may want to analyze their data with unsupported organisms, slim version of GO, novel functional annotation (e.g. GO via BlastGO or KEGG via KAAS), unsupported ontologies/pathways or customized annotations.

The `clusterProfiler` package provides `enricher()` function for hypergeometric test and `GSEA()` function for gene set enrichment analysis that are designed to accept user defined annotation. They accept two additional parameters `TERM2GENE` and `TERM2NAME`. As indicated in the parameter names, `TERM2GENE` is a data.frame with first column of term ID and second column of corresponding mapped gene and `TERM2NAME` is a `data.frame` with first column of term ID and second column of corresponding term name. `TERM2NAME` is optional.

## Input data

For over representation analysis, all we need is a gene vector, that is a vector of gene IDs. These gene IDs can be obtained by differential expression analysis (*e.g.* with the `DESeq2` package).

For gene set enrichment analysis, we need a ranked list of genes. `DOSE` provides an example dataset `geneList` which was derived from `R` package `breastCancerMAINZ` that contained 200 samples, including 29 samples in grade I, 136 samples in grade II and 35 samples in grade III. We computed the ratios of geometric means of grade III samples versus geometric means of grade I samples. Logarithm of these ratios (base 2) were stored in `geneList` dataset. If you want to prepare your own `geneList`, please refer to the [FAQ](#genelist).

We can load the sample data into R via:

```{r}
geneList <- readRDS("outputs/geneList.rds")
head(geneList)
```

Suppose we define fold change greater than 2 as DEGs:

```{r}
gene <- names(geneList)
head(gene)
```

## MSigDb analysis

[Molecular Signatures Database](http://software.broadinstitute.org/gsea/msigdb) is a collection of annotated gene sets. It contains 8 major collections:

-   H: hallmark gene sets
-   C1: positional gene sets
-   C2: curated gene sets
-   C3: motif gene sets
-   C4: computational gene sets
-   C5: GO gene sets
-   C6: oncogenic signatures
-   C7: immunologic signatures

Users can download [GMT files](www.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29) from [Broad Institute](http://software.broadinstitute.org/gsea/msigdb) and use the `read.gmt()` function to parse the file to be used in `enricher()` and `GSEA()`.

There is an R package, [msigdbr](https://cran.r-project.org/package=msigdbr), that already packed the MSigDB gene sets in tidy data format that can be used directly with `clusterProfiler` [@yu2012].

It supports several specices:

```{r}
library(msigdbr)
msigdbr_show_species()
# msigdbr_species()
```

We can retrieve all human gene sets:

```{r}
m_df <- msigdbr(species = "Rattus norvegicus")
head(m_df, 2) %>% as.data.frame
```

Or specific collection. Here we use C6, oncogenic gene sets as an example:

```{r}
m_t2g <- msigdbr(species = "Rattus norvegicus", category = "C6") %>% 
  dplyr::select(gs_name, gene_symbol)
head(m_t2g)
```

### MSigDb over-presentaton analysis {#msigdb-ora}

```{r}
em <- enricher(gene, TERM2GENE=m_t2g)
head(em)
```

### MSigDb gene set enrichment analysis {#msigdb-gsea}

In over-presentaton analysis, we use oncogenic gene sets (i.e. C6) to test whether the DE genes are involved in the process that leads to cancer. In this example, we will use the C3 category to test whether genes are up/down-regulated by sharing specific motif using GSEA approach.

```{r}
C3_t2g <- msigdbr(species = "Rattus norvegicus", category = "C3") %>% 
  dplyr::select(gs_name, gene_symbol)
head(C3_t2g)

em2 <- GSEA(geneList, TERM2GENE = C3_t2g, pvalueCutoff = 1)
head(em2)
```
